src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,94 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double INCREASING_ARRAY = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void growArray() {
+        if (size == elements.length) {
+            int newCapacity = (int) (elements.length * INCREASING_ARRAY);
+            T[] newElements = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, elements.length);
+            elements = newElements;
+        }
+    }
+
+    private void validateIndex(int index, boolean forAdd) {
+        if (index < 0 || index >= (forAdd ? size + 1 : size)) {
+            throw new ArrayListIndexOutOfBoundsException("Index:  " + index + " Size: " + size);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        growArray();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndex(index, true);
+        growArray();
```

Comment:
It's good that you're calling `growArray` before adding a new element to ensure there's enough space. However, the `growArray` method will always check if the array needs to grow even if it's not necessary. Consider checking if the array needs to grow inside the `add` and `addAll` methods before calling `growArray` to avoid unnecessary method calls.